dbuf/dbuf/dbuf.py

```diff
@@ -0,0 +1,287 @@
+            web.post('/s', self.create),
+            web.post('/s/{session}', self.post),
+            web.post('/s/{session}/get', self.get),
+            web.post('/s/{session}/getmany', self.getmany),
+            web.delete('/s/{session}', self.delete),
+            web.post('/w', self.post_worker),
+            web.get('/w', self.get_workers),
+            web.get('/healthcheck', self.healthcheck),
+        ])
+        self.app.on_cleanup.append(self.cleanup)
+
+    @staticmethod
+    async def serve(name, bufsize, data_dir, leader, binding_host='0.0.0.0', port=5000):
+        aiofiles = af.AIOFiles()
+        dbuf = await Sessions.make(bufsize, aiofiles, f'{data_dir}/{port}')
+        server = Server(name, binding_host, port, leader, dbuf, aiofiles)
+        try:
+            prefixed_app = server.deploy_config.prefix_application(server.app, server.name)
+            runner = web.AppRunner(prefixed_app)
+            await runner.setup()
+            site = web.TCPSite(runner, host=server.binding_host, port=server.port)
+            await site.start()
+            log.info(f'server {server.name} bound to {server.binding_host}:{server.port}')
+            if server.leader != server.name:
+                async def join_cluster():
+                    async with ah.ClientSession(raise_for_status=True,
+                                                timeout=ah.ClientTimeout(total=60)) as cs:
+                        async with cs.post(f'{server.leader_url}/w', data=server.name) as resp:
+                            assert resp.status == 200
+                            await resp.text()
+                await retry_forever(
+                    join_cluster,
+                    lambda exc: f'could not join cluster with leader {server.leader} at {server.leader_url} due to {exc}')
+            while True:
+                await asyncio.sleep(1 << 16)
+        finally:
+            await runner.cleanup()
+
+    def session(self, request):
+        session_id = int(request.match_info['session'])
+        session = self.dbuf.sessions.get(session_id)
+        if session is None:
+            raise web.HTTPNotFound(text=f'{session_id} not found')
+        return session
+
+    async def create(self, request):
+        session_id = await self.dbuf.new_session()
+        async with ah.ClientSession(raise_for_status=True,
+                                    timeout=ah.ClientTimeout(total=60)) as cs:
+            async def call(worker):
+                worker_url = self.deploy_config.base_url(worker)
+                async with cs.post(f'{worker_url}/s') as resp:
+                    assert resp.status == 200
+                    text = resp.text()
+                    assert await text == f'{session_id}', f'{text}, {session_id}'
+                    log.info(f'successfully created shuffle on {worker} at {worker_url}')
+            log.info(f'getting shuffle lock for {session_id}')
+            async with self.shuffle_create_lock:
```

Comment:
What's this lock about?  I'm skeptical.

Reply:
It makes session creation sequential. I wanted to make sure if worker i says it created the session that it had created the session I'm currently trying to create. Without it, sessions race to create themselves, some returning before the eventual state of all workers having all sessions.

I could fix by having workers create a session if someone POSTs data in a session of which they're unaware. That also makes growing the cluster (not currently supported) simpler.

Reply:
The whole distributed part of distributed buffer is, as you can tell, a big hack.